cmds/exp/gosh/gosh_test.go

```diff
@@ -0,0 +1,318 @@
+		},
+		{
+			name: "double brackets",
+			pairs: []string{
+				"[[\n",
+				"> ",
+				"true ]]\n",
+				"$ ",
+			},
+		},
+		{
+			name: "logic or",
+			pairs: []string{
+				"echo foo ||\n",
+				"> ",
+				"echo bar\n",
+				"foo\n",
+			},
+		},
+		{
+			name: "pipe",
+			pairs: []string{
+				"echo foo |\n",
+				"> ",
+				"read var; echo $var\n",
+				"foo\n",
+			},
+		},
+		{
+			name: "delayed newline",
+			pairs: []string{
+				"echo foo",
+				"",
+				" bar\n",
+				"foo bar\n",
+			},
+		},
+		{
+			name: "escaped newline",
+			pairs: []string{
+				"echo\\\n",
+				"> ",
+				" foo\n",
+				"foo\n",
+			},
+		},
+		{
+			name: "echo foo with escaped newline",
+			pairs: []string{
+				"echo foo\\\n",
+				"> ",
+				"bar\n",
+				"foobar\n",
+			},
+		},
+		{
+			name: "utf8",
+			pairs: []string{
+				"echo 你好\n",
+				"你好\n$ ",
+			},
+		},
+		{
+			name: "exit 0",
+			pairs: []string{
+				"echo foo; exit 0; echo bar\n",
+				"foo\n",
+				"echo baz\n",
+				"",
+			},
+		},
+		{
+			name: "exit 1",
+			pairs: []string{
+				"echo foo; exit 1; echo bar\n",
+				"foo\n",
+				"echo baz\n",
+				"",
+			},
+			wantErr: "exit status 1",
+		},
+		{
+			name: "no closing brace",
+			pairs: []string{
+				"(\n",
+				"> ",
+			},
+			wantErr: "1:1: reached EOF without matching ( with )",
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			inReader, inWriter := io.Pipe()
+			outReader, outWriter := io.Pipe()
+			runner, err := interp.New(interp.StdIO(inReader, outWriter, outWriter))
+			if err != nil {
+				t.Errorf("Failed creating runner: %v", err)
+			}
+			errc := make(chan error, 1)
+			go func() {
+				errc <- runInteractive(runner, inReader, outWriter)
+				if _, err := io.Copy(io.Discard, inReader); err != nil {
+					t.Errorf("Error discarding IO: %v", err)
+				}
+			}()
+
+			if err := readString(outReader, "$ "); err != nil {
+				t.Errorf("Failed reading string: %v", err)
+			}
+
+			for len(tt.pairs) > 0 {
+				if _, err := io.WriteString(inWriter, tt.pairs[0]); err != nil {
+					t.Errorf("Failed writing string: %v", err)
+				}
+				if err := readString(outReader, tt.pairs[1]); err != nil {
+					t.Errorf("Failed reading string: %v", err)
+				}
+
+				tt.pairs = tt.pairs[2:]
+			}
+
+			// Close the input pipe, so that the parser can stop
+			if err := inWriter.Close(); err != nil {
+				t.Errorf("Failed closing input pipe: %v", err)
+			}
+
+			// Once the inout pipe is closed, close the output pipe
+			// so that any remaining prompt writes get discarded
+			if err := outReader.Close(); err != nil {
+				t.Errorf("Failed closing output pipe: %v", err)
+			}
+
+			err = <-errc
+			if err != nil && tt.wantErr == "" {
+				t.Errorf("Unexpected error: %v", err)
+			} else if tt.wantErr != "" && fmt.Sprint(err) != tt.wantErr {
+				t.Errorf("Want error %q, got: %v", tt.wantErr, err)
+			}
+		})
+	}
+}
+
+func TestRunInteractiveTabCompletion(t *testing.T) {
+	// The function which should be tested here hardcodes
```

Comment:
Could you test it by passing a test-only completerFunc?

Reply:
I think it wasn't the completerFunc but rather the prompt.Input function that hardcodes device nodes internally. So I would have to create a test-only Input function in this case. That would require quite some refactoring but I could do so.